source/slang/syntax-base-defs.h

```diff
@@ -119,7 +119,7 @@ RAW(
 protected:
     virtual bool EqualsImpl(Type * type) = 0;
 
-    virtual Type* CreateCanonicalType() = 0;
+    virtual RefPtr<Type> CreateCanonicalType() = 0;
     Type* canonicalType = nullptr;
     RefPtr<Type> canonicalTypeRefPtr;
```

Comment:
I honestly don't remember when we added `canonicalTypeRefPtr` here, but I guess it was before this change.

In the past the challenge was that sometimes a type would be its own canonical type (because it is already canonical), and so a strong reference there would create a cycle. I suppose the answer is that the code handles this case by setting `canonicalType` to `this` and `canonicalTypeRefPtr` to `nullptr`...